src/scope.rs

```diff
@@ -1,42 +1,67 @@
-use std::collections::HashMap;
+use std::{cell::RefCell, collections::HashMap, rc::Rc};
 
 use crate::{err::InterpError, instruction::StackValue};
 
 #[derive(Debug)]
 pub struct Scope {
-    pub var_store: HashMap<String, StackValue>,
+    var_store: HashMap<String, StackValue>,
+    outter_scope: Option<Rc<RefCell<Scope>>>,
 }
 
 impl Scope {
     pub fn new() -> Self {
         Scope {
             var_store: HashMap::new(),
+            outter_scope: None,
         }
     }
 
-    pub fn get_var(&self, id: &String) -> Result<&StackValue, InterpError> {
-        self.var_store
-            .get(id)
-            .ok_or(InterpError::VariableNotFound(id.to_string()))
+    pub fn get_var_store_len(&self) -> usize {
+        self.var_store.len()
     }
 
-    pub fn set_var(&mut self, id: String, val: StackValue) -> Option<StackValue> {
-        self.var_store.insert(id, val)
-    }
-
-    pub fn from_scope(other: &Scope) -> Self {
-        let mut scope = Scope {
+    pub fn from_scope(outer_scope: Rc<RefCell<Scope>>) -> Self {
+        let scope = Scope {
             var_store: HashMap::new(),
+            outter_scope: Some(outer_scope),
         };
-        for (k, v) in other.var_store.iter() {
-            scope.set_var(k.to_string(), *v);
-        }
         return scope;
     }
 
-    pub fn assign(&mut self, kv: HashMap<&String, &StackValue>) {
-        for kv in kv.iter() {
-            self.set_var(kv.0.to_string(), **kv.1);
+    pub fn dec_var(&mut self, id: String) -> Option<StackValue> {
+        self.var_store.insert(id, StackValue::Uninitialised)
+    }
+
+    pub fn set_var(&mut self, id: String, val: StackValue) -> Result<StackValue, InterpError> {
+        let var = self.var_store.get(&id.clone());
+        match var {
+            Some(..) => {
+                self.var_store.insert(id, val);
+                Ok(val)
+            }
+            None => {
+                if let Some(out) = &mut self.outter_scope {
+                    return out.borrow_mut().set_var(id, val);
```

Comment:
This would be higher performance without recursion. I leave it up to you if you want to fix that or not.

Reply:
Good point.
I prefer to leave performance improvements for later, I think it should be driven by tests/benchmarking.

Reply:
Wise.

Did you want to add a comment or an entry in your todo list?

Reply:
Yes! added :point_right: 44f32d54371cd4aa3ec860838e84c8c599eb9b18